### ISSUE TITLE
failure to rematerialize 32-bit zero

### ISSUE THREAD

**Issue by matsc-at-sics-se** [matsc-at-sics-se/unison#46]
[prolog_datime6.mir.zip](https://github.com/matsc-at-sics-se/unison/files/2096953/prolog_datime6.mir.zip)
```
uni import -f prolog_datime6 prolog_datime6.mir -o prolog_datime6.uni --target=X86 --explicitcallreg --maxblocksize=25 --goal=speed --copyremat --lint
uni linearize prolog_datime6.uni -o prolog_datime6.lssa.uni --target=X86 --lint
uni extend prolog_datime6.lssa.uni -o prolog_datime6.ext.uni --target=X86 --lint
uni augment prolog_datime6.ext.uni -o prolog_datime6.alt.uni --target=X86 --copyremat --lint
uni model prolog_datime6.alt.uni -o prolog_datime6.json --target=X86 
gecode-presolver  prolog_datime6.json | aeson-pretty > prolog_datime6.ext.json
minizinc-solver -o prolog_datime6.out.json --topdown --chuffed --no-diffn --no-member --free --rnd --to 180 prolog_datime6.ext.json
* generated prolog_datime6.dzn in 70 msec
+ /home/matsc/unison/src/solvers/multi_backend/common/outfilter.pl prolog_datime6.out.json.last -
+ mzn-chuffed -a -s --fzn-flag -f --fzn-flag --rnd-seed --fzn-flag 123456 --fzn-flag --time-out --fzn-flag 180 -D good_cumulative=true -D good_diffn=false -D good_member=false prolog_datime6.mzn prolog_datime6.dzn
+ aeson-pretty
"cost": 15
15,1152,299238,1530,14,24,24282,1,0.42,0.02
==========

real	0m9.315s
user	0m9.120s
sys	0m0.124s
uni normalize prolog_datime6.asm.mir -o prolog_datime6.llvm.mir --target=X86 
uni export prolog_datime6.alt.uni --basefile=prolog_datime6.llvm.mir --solfile=prolog_datime6.out.json -o prolog_datime6.unison.mir --target=X86 
```
*.llvm.mir:
```
    PUSH64r %rbx, implicit-def %rsp, implicit %rsp
    %rbx = MOV64rr %rdi
    %edi = XOR32rr %edi, %edi, implicit-def %eflags
    CALL64pcrel32 @time, implicit %rsp
    %edx = XOR32rr %edx, %edx, implicit-def %eflags
    %rdi = MOV64rr %rbx
    %rsi = MOV64rr %rax
    %rbx = POP64r implicit-def %rsp, implicit %rsp
    TAILJMPd64 @datime_helper6, implicit %rsp
```
*.unison.mir:
```
    PUSH64r %rbx, implicit-def %rsp, implicit %rsp
    %rsp = SUB64ri32 %rsp, 16, implicit-def %eflags
    MOV64mr %rsp, 1, _, 8, _, %rdi
    %ebx = XOR32rr %ebx, %ebx, implicit-def %eflags
    %rdi = MOV64rr %rbx
    CALL64pcrel32 @time, implicit %rsp
    %edx = MOV32rr %ebx
    %rsi = MOV64rr %rax
    %rdi = MOV64rm %rsp, 1, _, 8, _
    %rsp = ADD64ri32 %rsp, 16, implicit-def %eflags
    %rbx = POP64r implicit-def %rsp, implicit %rsp
    TAILJMPd64 @datime_helper6, implicit %rsp
```

**Comment by matsc-at-sics-se** [comment]
Turns out to be an instance of #21! Due to:
```
[...]
    o12: [p28{-, t19}] <- {-, MOVE32, LOAD32, MOV32r0_remat} [p27{-, t16, t17}] (reads: [control,rsp], writes: [eflags], remat-origin: o9)
[...]
    o17: [p37{t23}:rax,p38{t24}:rcx_rdx,p39{t25}:rsi_rdi,p40{t26}:r8_r9,p41{t27}:r10_r11,p42{t28}:ymm0,p43{t29}:ymm1_15] <- (fun) [p36{t20, t21, t22}:rdi] (writes: [pc,eflags], call: o16)
[...]
    o22: [p55{-, t32}] <- {-, MOVE32, LOAD32, MOV32r0_remat} [p54{-, t16, t17}] (reads: [control,rsp], writes: [eflags], remat-origin: o9)
[...]
```
which (1) prevents `o12` from being scheduled after `o17`, and (2) prevents `o22` from playing the role of `o12`, because `o12` and `o22` are interchangeable.

**Comment by matsc-at-sics-se** [comment]
Solved by commit `7069b713b6f23899bf7a5ac5e8a4c647f1ed5919`.
